sdk/rust/oak_utils/src/lib.rs

```diff
@@ -41,31 +42,114 @@ pub fn run_protoc_rust(args: protoc_rust::Args) -> io::Result<()> {
     protoc_rust::run(temp_args)?;
 
     // Copy changed Rust `proto` files to the `out_path`.
-    copy_changed_files(temp_path, out_path)?;
+    update_files(out_path, temp_path)?;
 
     Ok(())
 }
 
 // Generates Rust `proto` files in a temporary directory using `run_protoc_rust`,
 // checks previously generated files and updates them if their contents have changed.
 pub fn run_protoc_rust_grpc(args: protoc_rust_grpc::Args) -> io::Result<()> {
-    let out_path = Path::new(args.out_dir);
+    let out_dir_path = Path::new(args.out_dir);
 
     // Create a temporary directory.
     let temp_dir = tempfile::tempdir()?;
-    let temp_path = temp_dir.path();
+    let temp_dir_path = temp_dir.path();
 
     // Generate Rust `grpc` files in the temporary directory.
     let mut temp_args = args;
-    temp_args.out_dir = temp_path.to_str().expect("Temporary path error");
+    temp_args.out_dir = temp_dir_path.to_str().expect("Temporary path error");
     protoc_rust_grpc::run(temp_args)?;
 
     // Copy changed Rust `grpc` files to the `out_path`.
-    copy_changed_files(temp_path, out_path)?;
+    update_files(out_dir_path, temp_dir_path)?;
 
     Ok(())
 }
 
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct FileDiff {
+    filename: String,
+    old_content: Option<String>,
+    new_content: Option<String>,
+}
+
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum FileAction {
+    Ignore,
+    Remove,
+    Update { new_content: String },
+}
+
+impl FileDiff {
+    // Returns a file action based on whether file contents were changed or not.
+    fn action(&self) -> FileAction {
+        match (self.old_content.as_ref(), self.new_content.as_ref()) {
+            (Some(old_content), Some(new_content)) => {
+                if old_content != new_content {
+                    FileAction::Update {
+                        new_content: new_content.to_string(),
+                    }
+                } else {
+                    FileAction::Ignore
+                }
+            }
+            (Some(_), None) => {
+                // `mod.rs` file is not generated by `protoc_rust` and `protoc_rust_grpc`.
+                if "mod.rs" != self.filename {
```

Comment:
These conditions are hard to follow because they are all negated, I think they would read more nicely if they were switched to be positive (e.g. `if old_content == new_content` and `if self.filename == "mod.rs"`)